content/blog/2018-06-07-when-to-use-derived-state.md

```diff
@@ -0,0 +1,331 @@
+The parent form component could then [use a `ref` to call this method](/docs/glossary.html#refs).
+
+> Note
+>
+> Refs can be useful when interfacing with imperative third party APIs or in certain, limited use cases like the one mentioned above, but generally we recommend you use them sparingly.
+
+## What about memoization?
+
+Another thing we've seen derived state used for is memoization. This isn't necessarily bad, but there are reasons you may want to avoid it. Hopefully the above examples illustrate that there is a certain amount of complexity inherent in using `getDerivedStateFromProps`. This complexity increases with each derived property. For example, if we add a second derived field to our component state then our implementation would need to separately track changes to both:
+
+```js
+class Example extends Component {
+  state = {};
+
+  static getDerivedStateFromProps(props, state) {
+    const updatedState = {};
+    if (props.bar !== state.prevBarFromProps) {
+      updatedState.prevBarFromProps = props.bar;
+      updatedState.derivedBar = deriveBar(props.bar);
+    }
+    if (props.foo !== state.prevFooFromProps) {
+      updatedState.prevFooFromProps = props.foo;
+      updatedState.derivedFoo = deriveFoo(props.foo);
+    }
+    // Other derived fields would go here...
+    return updatedState;
+  }
+
+  // ...
+}
+```
+
+Although the above pattern _works_, it isn't what `getDerivedStateFromProps` is meant to be used for. This component isn't reacting to **changes in props**. It's just computing values based on the **current props**, and there is a simpler way to do this that doesn't event require state.
+
+Let's look at another example to illustrate this. Here is a component that accepts an array of items as `props.list` and allows a user to filter the array by entering text. We could use `getDerivedStateFromProps` for this purpose:
+
+```js
+class Example extends Component {
+  state = {
+    filterText: "",
+    prevFilterText: ""
+  };
+
+  static getDerivedStateFromProps(props, state) {
+    // Re-run the filter whenever the list array or filter text change:
+    if (
+      props.list !== state.prevPropsList ||
+      state.prevFilterText !== state.filterText
+    ) {
+      // Note that to do this, state needs to explicitly track both:
+      // prevPropsList - so we can detect a change in props.list
+      // prevFilterText - so we can detect a change in state.filterText
+      return {
+        prevPropsList: props.list,
+        prevFilterText: state.filterText,
+        filteredList: props.list.filter(item => item.text.includes(state.filterText))
+      };
+    }
+    return null;
+  }
+
+  render() {
+    return (
+      <Fragment>
+        <input onChange={this.handleChange} value={this.state.filterText} />
+        <ul>{this.state.filteredList.map(item => <li key={item.id}>{item.text}</li>)}</ul>
+      </Fragment>
+    );
+  }
+
+  handleChange = event => this.setState({ filterText: event.target.value });
+}
+```
+
```

Comment:
I think this example would benefit from one intermediate simpler "after" version that just makes it a pure component, and does the calculation in render. I think we need to push harder against the preconception that you have to move out all the calculation from the render method. A lot of this comes from experience with Ember/Vue and not from any real performance concern.

Then we can say "but what if the component takes some other often-changing prop? Then memoization might be beneficial" and go into the example below.

Reply:
Oh, snap. I think I missed this one comment. I'm going to step away for a few minutes, and then I'll re-read this section with your comment in mind.

Reply:
I like this suggestion. I've pushed a change.